File: Homotopy.m
function [EV,EVT] = Homotopy(A,a,E,X)
%This function computes the eigenvalues and eigenvectors of a symmetrical
%tridiagonal matrix.
disp(nargin);
if nargin < 1
    error('Not enough arguments!');
end
global k n OO DD DDD
if nargin == 2
k = a;
[~,n] = size(A);
I = eye(n);
D = zeros(n,n);

%DD and OO are the diagonal and off diagonal elements of A
DD = zeros(n,1);
OO = zeros(n,1);

for i = 1:n
    DD(i) = A(i,i);
end

for i = 1:n
    D(i,i) = DD(i);
end

%d = diag(D);

OO(1) = 0;

for i = 2:n
    OO(i) = A(i,i-1);
end

for i = 1:n
    DDD(i) = DD(i)-OO(i);
end

%Compute the 1st 2nd and 3rd eigenvalue derivatives.
Z = DD(k);
XT = zeros(n,1);
XT(k) = 1;
alpha = 1;

Z1 = XT'*(A-D)*XT;
XT1 = pinv(Z*I-A)*(A-D)*XT;
Z2 = -2*(XT1')*XT1-((XT1')*XT1);
XT2 = 2*(A-D)*XT1+(alpha*Z1*XT1);
Z3 = -3*(XT1')*XT2-(3*(XT1')*XT2);

%NS is the number of steps
%H is the step size
%T is the homotopy paramemter
%PE is the predicted eigenvalue
NS = 0; H = 1; PT = 0; T = 1; X = zeros(n,1);

for i = 1:n
    X(i) = XT(i);
end

PE = Z+(H*Z1)+(H^2/2)*Z2+(H^3)*Z3;
disp('THE PREDICTED EIGENVALUE IS: ')
disp(PE);
disp('CALLING THE MAIN BLOCK');
mainblock(A,D,PE,XT,PT,T,NS,H,Z,Z1);
end

if nargin == 4
    EV = E;
    EVT = X;
    disp(EV);
    disp(EVT);
end

end

function [] = prediction(A,D,Z,PT,T,NS,H,X)
global n
disp('BEGINNING PREDICTION AGAIN');
alpha = 1;
I = eye(n);

XT = X;
Z1 = XT'*(A-D)*XT;
XT1 = pinv(Z*I-A)*(A-D)*XT;
Z2 = -2*(XT1')*XT1-((XT1')*XT1);
XT2 = 2*(A-D)*XT1+(alpha*Z1*XT1);
Z3 = -3*(XT1')*XT2-(3*(XT1')*XT2);

PE = Z+(H*Z1)+(H^2/2)*Z2+(H^3)*Z3;

mainblock(A,D,PE,XT,PT,T,NS,H,Z,Z1);

end


function [] = mainblock(A,D,PE,XT,PT,T,NS,H,Z,Z1)

global n k OO DD DDD 

%_________________________________________________________________________%
%Set the error tolerances
SUMA = 0;
SUMB = 0;

for i = 1:n
    SUMA=SUMA+abs(DD(i));
    SUMB=SUMB+abs(OO(i));
end

%NORM is defined as the sum of the sum of diagonal and off diagonal elements 
NORM = SUMA + SUMB;

EPS = eps*n*NORM;

if T == 1
    EPS1 = EPS;
else
    dt = norm(pinv(eye(n)-D));
    arr = [dt*sqrt(EPS),EPS];
    EPS1 = max(arr);
end

EPS2 = EPS; EPS3 = EPS;
%_________________________________________________________________________%

APP = PE;
X = XT;

F = zeros(n,1);
G = zeros(n,1);

for i = 1:n
    F(i) = D(i,i)+T*DDD(i);
    G(i) = T*OO(i);
end
disp('The value of T is: ');
disp(T);
t = T;

At = D + t*(A-D);

%Calculate the number of changes of the Sturm sequence
SC = Count(At,PE);
disp('THE NUMBER OF SIGN CHANGES IS: ')
disp(SC);

if (SC~=k) && (SC~=k-1)
    disp('BEGIN REDUCING STEP')
    reduceStep(A,D,Z,Z1,APP,NS,H,PT,XT);%GOTO 500
end

if SC == k
    KK = 1;
end

if SC == k-1
   KK = 0;
end

I = eye(n);
W = At-APP*I;
disp(W);

B = zeros(n,1);
Y = zeros(n,1);

for i = 1:n
    B(i) = F(i)-APP;
    Y(i) = X(i);
end
disp(B);
disp(Y);
%Perform inverse iteration with shift.
X = linsolve(W,Y);
SUM = 0;
disp(X);
for i = 1:n
    SUM = SUM+X(i)^2;
end
disp(SUM);
RES = 1/sqrt(SUM);

for i = 1:n
    X(i) = X(i)*RES;
end
disp(X);
pause
%If the norm of the residual vector is smaller than the tolerance,
%correction is stopped. If not, a new approximation for the eigenvalue is
%computed using the Rayleigh Quotient.
disp(RES);
disp(EPS1);
pause
if RES <= EPS1
    %GOTO 400
    disp('NORM OF THE RESIDUAL VECTOR IS SMALLER THAN TOLERANCE')
    val = APP-EPS1-EPS3;
    disp(val);
    sc = Count(At,val);
    disp('THE NUMBER OF SIGN CHANGES IS: ');
    disp(sc);
    if sc == k-1
        disp('BEGINNING COMPUTE AND PREDICT')
        computeANDpredict(A,D,NS,Z,Z1,APP,X,T,H);%GOTO 600
    end
    
    if sc > k-1
        disp('REDUCING STEP SIZE');
        reduceStep(A,D,Z,Z1,APP,NS,H,PT,XT);%GOTO 500
    end
    
    if sc < k-1
        val = APP+EPS1+EPS3;
        disp('COMPUTING SIGN CHANGE AGAIN');
        sc = Count(At,val);
        if sc >= k
            disp('BEGINNING COMPUTE AND PREDICT')
            computeANDpredict(A,D,NS,Z,Z1,APP,X,T,H)%GO TO 600
        end
    end
    
end

%PAPP is never actually used. 
%PAPP = APP;
HA = zeros(n,1);
disp('BEGINNING RAYLEIGH QUOTIENT ITERATION')
pause
for j = 1:10
    
    if j == 10
       disp('TOO MANY APPLICATIONS OF RQI')
       disp('BEGIN REDUCING STEP SIZE')
       reduceStep(A,D,Z,Z1,APP,NS,H,PT,XT);%GOTO 500
    end
    
    HA(1) = F(1)*X(1)+G(2)*X(2);
    for i = 2:n-1
        HA(i) = G(i)*X(i-1)+F(i)*X(i)+G(i+1)*X(i+1);
    end
    HA(n) = G(n)*X(n-1)+F(n)*X(n);

    APP = 0;
    for i = i:n
        APP = APP+(X(i)*HA(i));
    end
    disp('KK is: ');
    disp(KK);
    disp('APP is: ');
    disp(APP);
    disp('The predicted eigenvalue is: ');
    disp(PE);
    disp('EPS2 is: ');
    disp(EPS2);
    pause
    if (KK == 1) & (APP>(PE+EPS2))
        disp('BEGIN REDUCING STEP SIZE')
        reduceStep(A,D,Z,Z1,APP,NS,H,PT,XT);%GOTO 500
    end
    
    if (KK == 0) & (APP<(PE-EPS2))
        disp('BEGIN REDUCING STEP SIZE')
        reduceStep(A,D,Z,Z1,APP,NS,H,PT,XT);%GOTO 500
    end
end

end

function [] = reduceStep(A,D,Z,Z1,APP,NS,H,PT,XT)
%If the predicted or corrected eigenvalue was rejected, we reduce the step
%size and redo the predicition.
global n
    disp('REDUCING STEP SIZE');
    disp('The number of steps is: ')
    disp(NS);
    disp('The step size is: ')
    disp(H);
    disp('PT is: ');
    disp(PT);
    disp('The Vector is: ')
    disp(XT);
    H = H/2;
    T = PT+H;
    X = zeros(n,1);
    disp('The step size is now: ')
    disp(H);
    disp('The value of T is: ')
    disp(T);
    for i = 1:n
        X(i) = XT(i);
    end
    pause
    if(NS==0)
        disp('STEP 0. GOING BACK TO PREDICTION');
        prediction(A,D,Z,PT,T,NS,H,X)%GOTO 100
    else
        disp('STEP > 0. BEGINNING COMPUTE AND PREDICT');
        computeANDpredict(A,D,NS,Z,Z1,APP,X,T,H);
    end
    
end

function [] = computeANDpredict(A,D,NS,Z,Z1,APP,X,T,H)
%Recompute the eigenvalue derivative and then make a new prediction.
global n k OO DDD

NS = NS+1; OZ = Z; OZ1 = Z1; Z = APP;
XT = zeros(n,1);
HA = zeros(n,1);

for i = 1:n
    XT(i) = X(i);
end
disp('The eigenvector is: ');
disp(XT);
disp('THE VALUE OF T IS:')
disp(T);
pause
%If t == 1, we store the eigenpair
if T == 1
    [EV,EVT] = store(Z,XT);
    Homotopy(A,k,EV,EVT);
end
%If t ~= 1, we recompute the eigenvalue derivative.
if T < 1
    HA(1) = DDD(1)*X(1)+OO(2)*X(2);
    for i = 2:n-1
        HA(i) = OO(i)*X(i-1)+DDD(i)*X(i)+OO(i+1)*X(i+1);
    end
    HA(n) = OO(n)*X(n-1)+DDD(n)*X(n);
    Z1 = 0;
    for i = 1:n
        Z1 = Z1+X(i)*HA(i);
    end
end
disp('The new eigenvalue derivative is: ')
disp(Z1);
%Now we use hermite interpolation to obtain a new eigenvalue prediction.
PH = H; PT = T; H = 1-PT; T = 1;
Q = (1+H/PH);
QQ = Q*(H/PH);
PE = OZ+OZ1*(H+PH)+Q*((Z-OZ)-OZ*PH)+QQ*(PH*(Z1+OZ1)-2*(Z-OZ));
disp('RETURNING TO THE MAIN BLOCK');
disp(PE);
mainblock(A,D,PE,XT,PT,T,NS,H,Z,Z1);%GOTO 200

end

%if t = 1, we store the kth eigenpair of A 
function [EV,EVT] = store(Z,XT)
global n k
disp('BEGIN STORING THE EIGENPAIR');
EV = zeros(1,n);
EVT = zeros(n,n);

EV(k) = Z;

for i = 1:n 
    EVT(i) = XT(i); 
end

end

function [Count] = Count(At,x)
%COUNT finds the number of eigenvalues less than the predicted eigenvale.
%This ensures that we are on the correct eigenpath.
%At - the matrix at t = T
%x - the predicted eigenvalue
disp('COMPUTE SIGN CHANGES');
 Count = 0;
 d = 1;
 a = diag(At);
 b = diag(At,-1);
 [~,n] = size(At);
 
 for i = 1:n-1
     if i == 1
         d = a(i)-x;
     else
         d = a(i)-x-(b(i)^2)/d;
     end
     
     if d < 0
         Count = Count+1;
     end
 end

end

File: Homotopy.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <lapacke.h>
#include <cblas.h>
#include <time.h>

/*This program computes all of the eigenpairs of symmetric tridiagonal matricies using a Homotopy algorithm*/

//Global variables
const int N = 7500;
int k = 3240;

/*Function generates the type of matrix as described in Example 3*/
int Example3(double *A){
	int i,j;
	for(i = 0; i<N; i++){
		A[i*N+i]= (double) (i+1);
		for (j = 0; j<N-1; j++){
			A[(j*N+j)+1] = 1;
			A[(j*N+j)+N] = 1;
		}	
	}
	return 0;
}

/*Function generates an identity matrix of size n*/
int IdentityMat(double *ID){
	int i;
	double j=1.0;
	for (i = 0; i<N; i++){ID[(i*N)+i]=j;}
	return 0;
}


/*Initialize all of the vectors needed for computation*/
int Initialize(double *A, double *D, double *DD, double *DDD, double *OO){
	int i;
	for (i = 0; i<N; i++){
		DD[i] = A[i*N+i];
		D[i*N+i] = DD[i];
		DDD[i] = DD[i]-DD[i];
	}
	
	OO[0]=0.0;
	for (i = 1; i<N; i++){OO[i]=A[i*N+(i-1)];}
	return 0;
}

/*Compute the number of sign changes using Sturm Sequence method*/
int COUNT(double *DD, double *OO, double PE){
	int Count = 0,i;
	double d;
	d = 1.0;
	for (i = 0; i<N; i++){
		d = DD[i]-PE-((OO[i]*OO[i])/d);
		if (d<0.0){Count++;}
	}
	return Count;
}

/*This function reduces the step size, alters the homotopy parameter, and resets the eigenvector*/
int ReduceStep(double *A, double *D, double *DD, double *X, double *XT, double *DDD, double *OO, double H, double PT, double Z, double Z1, double APP, int NS){
	double T;
	H = H/2;
	T = PT+H;
	cblas_dcopy(N,XT,1,X,1);// X <- XT
	if (NS==0){
		double PE = Z+Z1*H;
		mainblock(A, D, DD, DDD, OO, X, XT, Z, Z1, H, T, PT, PE, NS);
	}
	else{
		printf("Recomputing the predicted eigenvalue...\n");
		Predict(A, D, DD, XT, X, DDD, OO, H, T, Z, Z1, APP, NS);
	}		
	return 0;	
}

int initialPrediction(int k){
	    //Matrices
		double *A = (double*)calloc(N*N,sizeof(double));
		double *D = (double*)calloc(N*N,sizeof(double));
		double *AD = (double*)calloc(N*N,sizeof(double));
		//Vectors	
		double *DD = (double*)calloc(N,sizeof(double));
		double *DDD = (double*)calloc(N,sizeof(double));
		double *OO = (double*)calloc(N,sizeof(double));
		double *X = (double*)calloc(N,sizeof(double));	
		double *XT = (double*)calloc(N,sizeof(double));
		double *XTprime = (double*)calloc(N,sizeof(double));
		//Other important variables
		double alpha, beta, H, PT, T, Z, Z1, PE;	
		int NS, i, j, incx, incy;
		//clock_t t; 
		Example3(A);
		Initialize(A,D,DD,DDD,OO);
		for (i = 0; i<N; i++){
			for (j = 0; j<N; j++){
				AD[i*N+j] = A[i*N+j]-D[i*N+j];		
			}
		}
	
		H = T = alpha = beta = 1.0;
		PT = 0.0;
		incx = incy = 1;	
		NS = 0;

		printf("Beginning Computation\n");

		Z=DD[k-1];
		XT[k-1]=1;
		cblas_dcopy(N,XT,incx,X,incx);//X <- XT
		//Compute first derivative
		printf("Computing Z1\n");
		cblas_dgemv(CblasRowMajor,CblasNoTrans,N,N,alpha,AD,N,XT,incx,beta,XTprime,incy); //XT' <- (A-D)*XT
		Z1 = cblas_ddot(N,XT,incx,XTprime,incx);//Z1 <- XT'*((A-D)*XT)
		//Predict eigenvalue with first order taylor method
		PE=Z+Z1*H;
		printf("Beginning Correction.\n");
		free(AD);
		free(XTprime);
		mainblock(A, D, DD, DDD, OO, X, XT, Z, Z1, H, T, PT, PE, NS);
	return 0;
}


/*This function computes the residual of the vector obtain from II*/
double computeRES(double* A, double* X, double APP){
	double *AX = (double*)calloc(N,sizeof(double));
	double *APPX = (double*)calloc(N,sizeof(double));
	double RES,alpha,beta;
	int incx,incy,i,j;
	alpha = beta = 1.0;
	incx = incy = 1;
	cblas_dcopy(N,X,incx,APPX,incx);//APPY <- Y
	cblas_dscal(N,-APP,APPX,incx);//APPY <- APP*Y
	cblas_dgemv(CblasRowMajor,CblasNoTrans,N,N,alpha,A,N,X,incx,beta,AX,incx);//AY <- A*Y
	cblas_daxpy(N,alpha,AX,incx,APPX,incx);
	RES = cblas_dnrm2(N,APPX,1);
	printf("The maximum residual is %.17e\n",RES);
	free(AX);
	free(APPX);
	return RES;
}
/*Function recomputes PE using Hermite Interpolation.*/
/*If T == 1, the eigenvalues are stored.*/
int Predict(double *A, double *D, double *DD, double *XT,double *X, double *DDD, double *OO, double H, double T, double Z, double Z1, double APP, int NS){
	double *HA = (double*)calloc(N,sizeof(double));	
	double OZ=Z, OZ1=Z1, Q, QQ, PH=H, PT=T,PE,RES,maxRES,ORT,maxORT;
	int i,j;

	H = 1-PT; T = 1; NS+=1; 
	
	cblas_dcopy(N,X,1,XT,1); //XT<-X
	
	if (T==1){
		printf("Storing the eigenvalue and eigenvector!\n");
		printf("%.15f\n",APP);
		if (k==1){
			maxRES = maxORT = 0.0;
			RES = maxRES;
			ORT = maxORT;
		}
		RES = computeRES(A,X,APP);
		if (RES>maxRES){maxRES = RES;}
		
		///////////////////////
		free(A);
		free(D);
		free(DD);
		free(DDD);
		free(OO);
		free(X);
		free(XT);
		///////////////////////
		k++;
		if (k>N){
			printf("The maximum residual is %.17e\n",maxRES);
			return 0;		
		}	
		return 0;	
	}
	else{
		//Recompute Z1
		HA[0]=DDD[0]*X[0]+OO[1]*X[1];
		for (i = 1; i<N-1; i++){HA[i]=OO[i]*X[i-1]+DDD[i]*X[i]+OO[i+1]*X[i+1];}
		HA[N-1]=OO[N-1]*X[N-2]+DDD[N-1]*X[N-1];
		//Recompute PE
		Q = pow(1+(PH/H),2);
		QQ = Q*(H/PH);
		PE=OZ+OZ1*(H+PH)+Q*((Z-OZ)-OZ1*PH)+QQ*(PH*(Z+OZ1)-2*(Z-OZ));

		mainblock(A, D, DD, DDD, OO, X, XT, Z, Z1, H, T, PT, PE, NS);
	}
	free(HA);
	return 0;
}


/*Approximate an eigenvector with Inverse Iteration*/
double * II(double *At, double *X, double APP, int j){
	printf("Beginning Inverse Iteration...\n");
	double *W = (double*)calloc(N*N,sizeof(double));
	double *ID = (double*)calloc(N*N,sizeof(double));
	double *DL = (double*)calloc(N-1,sizeof(double));
	double *D = (double*)calloc(N,sizeof(double));
	double *DU = (double*)calloc(N-1,sizeof(double));
	int *IPIV = (int*)malloc(N*sizeof(int));
	double norm=0, SUM = 0;	
	int i,l,NRHS=1,LDB=1,LDA=N, INFO;
	IdentityMat(ID);
	//W = At-APP*I
	for (i = 0; i<N; i++){
		for (l = 0; l<N; l++){
			W[i*N+l] = At[i*N+l] - APP*ID[i*N+l];		
		}	
	}
	//Get diagonal
	for (i = 0; i<N; i++){
		D[i] = W[i*N+i];	
	}
	//Get upper and lower diagonals
	for (i = 0; i<N-1; i++){
		DL[i] = W[(i*N+i)+1];
		DU[i] = DL[i];	
	}
	norm = cblas_dnrm2(N,X,1); //norm(X);
	cblas_dscal(N,1/norm,X,1); // X <- X*1/norm(X)
    printf("Solving the system of linear equations...\n"); 
    INFO = LAPACKE_dgtsv(LAPACK_ROW_MAJOR,N,NRHS,DL,D,DU,X,LDB);
	norm = cblas_dnrm2(N,X,1); //norm(X);
	cblas_dscal(N,1/norm,X,1); // X <- X*1/norm(X)
	free(W);
	free(ID);
	free(IPIV);
	printf("Ending Inverse Iteration\n");
	//printf("The new approximate eigenvector is: \n");
	//printVec(X);
	return X;
}

/*Approximate an eigenvalue with Rayleigh Quotient*/
double RQI(double *At, double *X, int j){
	double *u = (double*)calloc(N,sizeof(double));
	double norm,APP,SUM,alpha,beta;
	int incx,incy,i;
	norm = APP = SUM = 0;
	alpha = beta = 1.0;
	incx = incy = 1;
	printf("Beginning Rayleigh Quotient Iteration...\n");
	//Peform Rayleigh Quotient Iteration
	norm = cblas_dnrm2(N,X,1); //norm(X)
	cblas_dscal(N,1/norm,X,incx); //X <- X*1/norm(X);
	cblas_dgemv(CblasRowMajor,CblasNoTrans,N,N,alpha,At,N,X,incx,beta,u,incy); //u <- At*X
	APP = cblas_ddot(N,u,incx,X,incx); //APP <- X'*u
	printf("Ending Rayleigh Quotient Iteration...\n");
	printf("The new approximate eigenvalue is...\n");
	printf("APP = %.10e\n",APP);
	free(u);
	return APP;
}

/*The main block of the program. It corrects the prediction computed.*/
int mainblock(double *A, double *D, double *DD, double *DDD, double *OO, double *X, double *XT, double Z, double Z1, double H, double T, double PT, double PE, int NS){
	double eps, EPS, EPS1, EPS2, EPS3, NORM, SUMA, SUMB, APP, RES;
	int i,sc,KK,j;
	
	//Set the error tolerances for the program
	eps = 1.11E-16;
	EPS = EPS1 = EPS2 = EPS3 = NORM = SUMA = SUMB = RES = 0;
	for (i = 0; i<N; i++){
		SUMA+=abs(DD[i]);
		SUMB+=abs(OO[i]);
	}
	
	NORM = SUMA+SUMB;
	
	EPS=eps*N*NORM;
	EPS2=EPS3=EPS;

	if (T==1){EPS1=EPS;}
	else{EPS1=sqrt(EPS);}
	
	//////////////////////////////////////////////////

	double *At = (double*)calloc(N*N,sizeof(double));
	
	//Form At = D + T*(A-D);
	for (i=0; i<N; i++){
		for (j=0; j<N; j++){
			At[i*N+j] = D[i*N+j]+T*(A[i*N+j]-D[i*N+j]);		
		}
	}

	APP = PE;
	sc = COUNT(DD,OO,APP);
	//printf("The number of sign changes is %d\n",sc);
	
	if (sc!=k && sc!=k-1){ReduceStep(A,D,DD,X,XT,DDD,OO,H,PT,Z,Z1,APP,NS); return 0;}	
	if (sc==k){KK=1;}
	if (sc==k-1){KK=0;}
	
	for (i = 1; i<=10; i++){
		X = II(At,X,APP,1);
		//Compute Residual
		printf("Computing the residual...\n");
		RES = computeRES(A,X,APP);
		printf("Epsilon is: %e\n",EPS1);

		if (RES<=EPS1){
			free(At);
			sc = COUNT(DD,OO,APP-EPS1-EPS3);
			printf("The number of sign changes is %d\n",sc);
			if (sc==k-1){
				printf("Recomputing the predicted eigenvalue...\n");
				Predict(A,D,DD,XT,X,DDD,OO,H,T,Z,Z1,APP,NS);
				return 0;			
			}
			else if (sc>k-1){
				printf("Reducing the step size...\n");
				ReduceStep(A,D,DD,X,XT,DDD,OO,H,PT,Z,Z1,APP,NS);
				return 0;
			}
			else{
				sc = COUNT(DD,OO,APP+EPS1+EPS3);
				printf("The number of sign changes is %d\n",sc);
				if (sc>=k){
					printf("Recomputing the predicted eigenvalue...\n");
					Predict(A,D,DD,XT,X,DDD,OO,H,T,Z,Z1,APP,NS);
					return 0;				
				}
			}		
		}
		
		APP = RQI(At,X,1);
		if (KK==1 && APP>(PE+EPS2)){
			printf("Reducing the step size...\n");
			ReduceStep(A,D,DD,X,XT,DDD,OO,H,PT,Z,Z1,APP,NS);
			return 0;
		}
		
		if (KK==0 && APP<(PE-EPS2)){
			printf("Reducing the step size...\n");
			ReduceStep(A,D,DD,X,XT,DDD,OO,H,PT,Z,Z1,APP,NS);
			return 0;
		}
		//Reduce step if too many iterations of II/RQI
		if (i==10){
			printf("Reducing the step size...\n");
			ReduceStep(A,D,DD,X,XT,DDD,OO,H,PT,Z,Z1,APP,NS);
			return 0;
		}
	}
	return 0;		
}

/*This function prints a matrix*/
int printMat(double *A){
	int i,j;
	for (i=0; i<N; i++){
			for (j=0; j<N; j++){printf("%f\t",A[i*N+j]);}
			printf("\n");		
	}
	printf("\n");
}

int printVec(double *X){
	int i;
	for (i = 0; i<N; i++){printf("%.15f\n",X[i]);}
	return 0;	
}


//Main
int main(void){
	int i;
	clock_t begin, end;
	double time;
	begin = clock();	
	for (i = 1; i<=21; i++){	
		initialPrediction(k);
		printf("k = %d\n",k);	
	}
	end = clock();
	time = (double)(end-begin)/CLOCKS_PER_SEC;	
	printf("The total computation time is: %f\n",time);
	return 0;
}
